contracts/lottery/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+    }
+
+    /// Allows any participant with enough funds to buy a ticket.
+    ///
+    /// # Arguments
+    ///
+    /// - `env` - The environment for this contract.
+    /// - `by` - The address that is buying the ticket. Its enforced to match with the incoming transaction signatures.
+    /// - `ticket` - The selected numbers by the player
+    pub fn buy_ticket(env: Env, by: Address, ticket: Vec<u32>) -> Result<u32, Error> {
+        by.require_auth();
+
+        let storage = env.storage().persistent();
+
+        lottery_must_be_active(&storage)?;
+
+        let number_of_elements = storage.get::<_, u32>(&DataKey::NumberOfNumbers).unwrap();
+        let max_range = storage.get::<_, u32>(&DataKey::MaxRange).unwrap();
+
+        if ticket.len() != number_of_elements {
+            return Err(Error::NotEnoughOrTooManyNumbers);
+        }
+
+        //each number must be within (1, max_range) range
+        for number in ticket.iter() {
+            if number <= 0 || number > max_range {
+                return Err(Error::InvalidNumbers);
+            }
+        }
+
+        let price = storage.get::<_, i128>(&DataKey::TicketPrice).unwrap();
+        let token = storage.get::<_, Address>(&DataKey::Token).unwrap();
+        let token_client = token::Client::new(&env, &token);
+
+        if token_client.balance(&by) <= price {
+            return Err(Error::InsufficientFunds);
+        }
+
+        token_client.transfer(&by, &env.current_contract_address(), &price);
+
+        let mut tickets = storage
+            .get::<_, Map<Address, Vec<LotteryTicket>>>(&DataKey::Tickets)
+            .unwrap();
+
+        let mut player_selection = tickets.get(by.clone()).unwrap_or(vec![&env]);
+        player_selection.push_back(ticket);
+        tickets.set(by, player_selection);
+
+        storage.set(&DataKey::Tickets, &tickets);
+        Ok(tickets.values().len())
+    }
+
+    /// Returns actual pool balance. Can only be invoked by admin
+    ///
+    /// # Arguments
+    ///
+    /// - `env` - The environment for this contract.
+    pub fn get_pool_balance(env: Env) -> Result<i128, Error> {
+        let storage = env.storage().persistent();
+        let admin = storage.get::<_, Address>(&DataKey::Admin).unwrap();
+        admin.require_auth();
+
+        let token = storage.get::<_, Address>(&DataKey::Token).unwrap();
+        let token_client = token::Client::new(&env, &token);
+
+        Ok(token_client.balance(&env.current_contract_address()))
+    }
+
+    /// Returns results of a given lottery. If no results are available, or wrong lottery number
+    /// was given an error is returned
+    ///
+    /// # Returns
+    /// 
+    /// - Results of the lottery
+    /// 
+    /// # Arguments
+    ///
+    /// - `lottery_number` - Number of the lottery
+    pub fn check_lottery_results(env: Env, lottery_number: u32) -> Result<Vec<u32>, Error> {
+        let storage = env.storage().persistent();
+        let lottery_results_opt = storage
```

Comment:
This would remove the is_none, and the unwrap
```suggestion
        let lottery_results_opt = storage.get::<_, Map<u32, LotteryResult>>(&DataKey::LotteryResults).ok_or(Error::NoLotteryResultsAvailable)?;
```